// ngx-nestable/bundles/ngx-nestable.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/core', '@angular/common'], factory) :
	(factory((global['ngx-nestable'] = {}),global.ng.core,global.ng.common));
}(this, (function (exports,core,common) { 'use strict';

var _traverseChildren = function (tree, callback, parent) {
    if (parent === void 0) { parent = null; }
    for (var /** @type {?} */ i = 0; i < tree.length; i++) {
        var /** @type {?} */ item = tree[i];
        if (typeof item === 'undefined') {
            continue;
        }
        var /** @type {?} */ callbackResult = callback(item, parent);
        if (callbackResult) {
            break;
        }
        if (item.children) {
            _traverseChildren(item.children, callback, item);
        }
    }
};
var _insertAfter = function (newNode, referenceNode) {
    referenceNode.parentNode.insertBefore(newNode, referenceNode.nextSibling);
};
var _getParents = function (el, parentSelector) {
    if (parentSelector === void 0) { parentSelector = document.body; }
    var /** @type {?} */ parents = [];
    var /** @type {?} */ parentNode = el.parentNode;
    while (parentNode !== parentSelector) {
        var /** @type {?} */ o = parentNode;
        if (!parentNode) {
            break;
        }
        if (parentNode.tagName === parentSelector.tagName) {
            parents.push(o);
        }
        parentNode = o.parentNode;
    }
    parents.push(parentSelector); // Push that parentSelector you wanted to stop at
    return parents;
};
var _closest = function (el, selector) {
    var /** @type {?} */ matchesFn;
    // find vendor prefix
    ['matches', 'webkitMatchesSelector', 'mozMatchesSelector', 'msMatchesSelector', 'oMatchesSelector'].some(function (fn) {
        if (typeof document.body[fn] === 'function') {
            matchesFn = fn;
            return true;
        }
        return false;
    });
    var /** @type {?} */ parent;
    // traverse parents
    while (el) {
        parent = el.parentElement;
        if (parent === null) {
            break;
        }
        var /** @type {?} */ matches = parent[matchesFn](selector);
        if (parent && matches) {
            return parent;
        }
        el = parent;
    }
    return null;
};
var _offset = function (elem) {
    var /** @type {?} */ box = { top: 0, left: 0 };
    // BlackBerry 5, iOS 3 (original iPhone)
    if (typeof elem.getBoundingClientRect !== undefined) {
        box = elem.getBoundingClientRect();
    }
    return {
        top: box.top + (window.pageYOffset || elem.scrollTop) - (elem.clientTop || 0),
        left: box.left + (window.pageXOffset || elem.scrollLeft) - (elem.clientLeft || 0)
    };
};
var _findObjectInTree = function (array, id) {
    var /** @type {?} */ result = null;
    _traverseChildren(array, function (item) {
        if (item['$$id'] === Number.parseInt(id)) {
            result = item;
            return true;
        }
    });
    return result;
};
var REGISTER_HANDLE = 'NESTABLE_DRAG_HANDLE_REGISTER';
var DRAG_START = 'NESTABLE_DRAG_HANDLE_START';
var EXPAND_COLLAPSE = 'NESTABLE_EXPAND_COLLAPSE_EVENT';
var defaultSettings = ({
    listNodeName: 'ul',
    itemNodeName: 'li',
    rootClass: 'dd',
    listClass: 'dd-list',
    itemClass: 'dd-item',
    dragClass: 'dd-dragel',
    handleClass: 'dd-handle',
    collapsedClass: 'dd-collapsed',
    placeClass: 'dd-placeholder',
    group: 0,
    maxDepth: 5,
    threshold: 20,
    fixedDepth: false,
    exportCollapsed: true,
    disableDrag: false,
});
var mouse = {
    moving: 0,
    offsetX: 0,
    offsetY: 0,
    startX: 0,
    startY: 0,
    lastX: 0,
    lastY: 0,
    nowX: 0,
    nowY: 0,
    distX: 0,
    distY: 0,
    dirAx: 0,
    dirX: 0,
    dirY: 0,
    lastDirX: 0,
    lastDirY: 0,
    distAxX: 0,
    distAxY: 0
};
var PX = 'px';
var hasPointerEvents = (function () {
    var /** @type {?} */ el = document.createElement('div'), /** @type {?} */ docEl = document.documentElement;
    if (!('pointerEvents' in el.style)) {
        return false;
    }
    el.style.pointerEvents = 'auto';
    el.style.pointerEvents = 'x';
    docEl.appendChild(el);
    var /** @type {?} */ supports = window.getComputedStyle &&
        window.getComputedStyle(el, '').pointerEvents === 'auto';
    docEl.removeChild(el);
    return !!supports;
})();
var NestableComponent = (function () {
    /**
     * @param {?} ref
     * @param {?} renderer
     * @param {?} el
     * @param {?} zone
     */
    function NestableComponent(ref, renderer, el, zone) {
        this.ref = ref;
        this.renderer = renderer;
        this.el = el;
        this.zone = zone;
        this.listChange = new core.EventEmitter();
        this.drop = new core.EventEmitter();
        this.drag = new core.EventEmitter();
        this.disclosure = new core.EventEmitter();
        this.options = defaultSettings;
        this.disableDrag = false;
        this.dragRootEl = null;
        this.dragEl = null;
        this.dragModel = null;
        this.moving = false;
        /**
         * Dragged element contains children, and those children contain other children and so on...
         * This property gives you the number of generations contained within the dragging item.
         */
        this.dragDepth = 0;
        /**
         * The depth of dragging item relative to element root (ngx-nestable)
         */
        this.relativeDepth = 0;
        this.hasNewRoot = false;
        this.pointEl = null;
        this.items = [];
        this._componentActive = false;
        this._mouse = Object.assign({}, mouse);
        this._list = [];
        this._itemId = 0;
        this._registerHandleDirective = false;
    }
    Object.defineProperty(NestableComponent.prototype, "list", {
        /**
         * @return {?}
         */
        get: function () {
            return this._list;
        },
        /**
         * @param {?} list
         * @return {?}
         */
        set: function (list) {
            this._list = list;
            this._generateItemIds();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    NestableComponent.prototype.ngOnInit = function () {
        // set/extend default options
        this._componentActive = true;
        var /** @type {?} */ optionKeys = Object.keys(defaultSettings);
        for (var _i = 0, optionKeys_1 = optionKeys; _i < optionKeys_1.length; _i++) {
            var key = optionKeys_1[_i];
            if (typeof this.options[key] === 'undefined') {
                this.options[key] = defaultSettings[key];
            }
        }
        this._generateItemIds();
        this._generateItemExpanded();
        this._createHandleListener();
    };
    /**
     * @return {?}
     */
    NestableComponent.prototype.ngOnDestroy = function () { };
    /**
     * @return {?}
     */
    NestableComponent.prototype._generateItemIds = function () {
        var _this = this;
        _traverseChildren(this._list, function (item) {
            item['$$id'] = _this._itemId++;
        });
    };
    /**
     * @return {?}
     */
    NestableComponent.prototype._generateItemExpanded = function () {
        _traverseChildren(this._list, function (item) {
            if (typeof item.expanded === 'undefined') {
                item['$$expanded'] = true;
            }
            else {
                item['$$expanded'] = item.expanded;
            }
        });
    };
    /**
     * @return {?}
     */
    NestableComponent.prototype._createHandleListener = function () {
        var _this = this;
        this.renderer.listen(this.el.nativeElement, REGISTER_HANDLE, function () {
            _this._registerHandleDirective = true;
        });
        this.renderer.listen(this.el.nativeElement, DRAG_START, function (data) {
            _this.dragStart(data.detail.event, data.detail.param.item, data.detail.param.parentList);
        });
        this.renderer.listen(this.el.nativeElement, EXPAND_COLLAPSE, function (data) {
            _this.disclosure.emit({
                item: data.detail.item,
                expanded: data.detail.item['$$expanded']
            });
        });
    };
    /**
     * @param {?} event
     * @param {?} dragItem
     * @return {?}
     */
    NestableComponent.prototype._createDragClone = function (event, dragItem) {
        this._mouseStart(event, dragItem);
        if (!this._registerHandleDirective) {
            this._mouse.offsetY = dragItem.nextElementSibling
                ? dragItem.nextElementSibling.clientHeight / 2
                : dragItem.clientHeight / 2;
        }
        // create drag clone
        this.dragEl = document.createElement(this.options.listNodeName);
        document.body.appendChild(this.dragEl);
        this.renderer.addClass(this.dragEl, this.options.dragClass);
        // add drag clone to body and set css
        this.renderer.setStyle(this.dragEl, 'left', event.pageX - this._mouse.offsetX + PX);
        this.renderer.setStyle(this.dragEl, 'top', event.pageY - this._mouse.offsetY + PX);
        this.renderer.setStyle(this.dragEl, 'position', 'absolute');
        this.renderer.setStyle(this.dragEl, 'z-index', 9999);
        this.renderer.setStyle(this.dragEl, 'pointer-events', 'none');
    };
    /**
     * @param {?} event
     * @param {?} dragItem
     * @return {?}
     */
    NestableComponent.prototype._createPlaceholder = function (event, dragItem) {
        this._placeholder = document.createElement('div');
        this._placeholder.classList.add(this.options.placeClass);
        _insertAfter(this._placeholder, dragItem);
        dragItem.parentNode.removeChild(dragItem);
        this.dragEl.appendChild(dragItem);
        this.dragRootEl = dragItem;
    };
    /**
     * Sets depth proerties (relative and drag)
     * @return {?}
     */
    NestableComponent.prototype._calculateDepth = function () {
        // total depth of dragging item
        var /** @type {?} */ depth;
        var /** @type {?} */ items = this.dragEl.querySelectorAll(this.options.itemNodeName);
        for (var /** @type {?} */ i = 0; i < items.length; i++) {
            depth = _getParents(items[i], this.dragEl).length;
            if (depth > this.dragDepth) {
                this.dragDepth = depth;
            }
        }
        // depth relative to root
        this.relativeDepth = _getParents(this._placeholder, this.el.nativeElement.querySelector(this.options.listNodeName)).length;
    };
    /**
     * @param {?} event
     * @param {?} dragItem
     * @return {?}
     */
    NestableComponent.prototype._mouseStart = function (event, dragItem) {
        this._mouse.offsetX = event.pageX - _offset(dragItem).left;
        this._mouse.offsetY = event.pageY - _offset(dragItem).top;
        this._mouse.startX = this._mouse.lastX = event.pageX;
        this._mouse.startY = this._mouse.lastY = event.pageY;
    };
    /**
     * @param {?} event
     * @return {?}
     */
    NestableComponent.prototype._mouseUpdate = function (event) {
        // mouse position last events
        this._mouse.lastX = this._mouse.nowX;
        this._mouse.lastY = this._mouse.nowY;
        // mouse position this events
        this._mouse.nowX = event.pageX;
        this._mouse.nowY = event.pageY;
        // distance mouse moved between events
        this._mouse.distX = this._mouse.nowX - this._mouse.lastX;
        this._mouse.distY = this._mouse.nowY - this._mouse.lastY;
        // direction mouse was moving
        this._mouse.lastDirX = this._mouse.dirX;
        this._mouse.lastDirY = this._mouse.dirY;
        // direction mouse is now moving (on both axis)
        this._mouse.dirX =
            this._mouse.distX === 0 ? 0 : this._mouse.distX > 0 ? 1 : -1;
        this._mouse.dirY =
            this._mouse.distY === 0 ? 0 : this._mouse.distY > 0 ? 1 : -1;
    };
    /**
     * @return {?}
     */
    NestableComponent.prototype._showMasks = function () {
        var /** @type {?} */ masks = this.el.nativeElement.getElementsByClassName('nestable-item-mask');
        for (var /** @type {?} */ i = 0; i < masks.length; i++) {
            masks[i].style.display = 'block';
        }
    };
    /**
     * @return {?}
     */
    NestableComponent.prototype._hideMasks = function () {
        var /** @type {?} */ masks = this.el.nativeElement.getElementsByClassName('nestable-item-mask');
        for (var /** @type {?} */ i = 0; i < masks.length; i++) {
            masks[i].style.display = 'none';
        }
    };
    /**
     * calc mouse traverse distance on axis
     * @param {?} m - mouse
     * @return {?}
     */
    NestableComponent.prototype._calcMouseDistance = function (m) {
        m.distAxX += Math.abs(m.distX);
        if (m.dirX !== 0 && m.dirX !== m.lastDirX) {
            m.distAxX = 0;
        }
        m.distAxY += Math.abs(m.distY);
        if (m.dirY !== 0 && m.dirY !== m.lastDirY) {
            m.distAxY = 0;
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    NestableComponent.prototype._move = function (event) {
        var /** @type {?} */ depth, /** @type {?} */ list;
        var /** @type {?} */ dragRect = this.dragEl.getBoundingClientRect();
        this.renderer.setStyle(this.dragEl, 'left', event.pageX - this._mouse.offsetX + PX);
        this.renderer.setStyle(this.dragEl, 'top', event.pageY - this._mouse.offsetY + PX);
        this._mouseUpdate(event);
        // axis mouse is now moving on
        var /** @type {?} */ newAx = Math.abs(this._mouse.distX) > Math.abs(this._mouse.distY) ? 1 : 0;
        // do nothing on first move
        if (!this._mouse.moving) {
            this._mouse.dirAx = newAx;
            this._mouse.moving = 1;
            return;
        }
        // calc distance moved on this axis (and direction)
        if (this._mouse.dirAx !== newAx) {
            this._mouse.distAxX = 0;
            this._mouse.distAxY = 0;
        }
        else {
            this._calcMouseDistance(this._mouse);
        }
        this._mouse.dirAx = newAx;
        // find list item under cursor
        if (!hasPointerEvents) {
            this.dragEl.style.visibility = 'hidden';
        }
        var /** @type {?} */ pointEl = document.elementFromPoint(event.pageX - document.body.scrollLeft, event.pageY - (window.pageYOffset || document.documentElement.scrollTop));
        if (!hasPointerEvents) {
            this.dragEl.style.visibility = 'visible';
        }
        if (pointEl &&
            (pointEl.classList.contains('nestable-item-mask') ||
                pointEl.classList.contains(this.options.placeClass))) {
            this.pointEl = pointEl.parentElement.parentElement;
        }
        else {
            return;
        }
        /**
         * move horizontal
         */
        if (!this.options.fixedDepth &&
            this._mouse.dirAx &&
            this._mouse.distAxX >= this.options.threshold) {
            // reset move distance on x-axis for new phase
            this._mouse.distAxX = 0;
            var /** @type {?} */ previous = this._placeholder.previousElementSibling;
            // increase horizontal level if previous sibling exists, is not collapsed, and can have children
            if (this._mouse.distX > 0 && previous) {
                list = previous.querySelectorAll(this.options.listNodeName);
                list = list[list.length - 1];
                // check if depth limit has reached
                depth = _getParents(this._placeholder, this.el.nativeElement.querySelector(this.options.listNodeName)).length;
                if (depth + this.dragDepth <= this.options.maxDepth) {
                    // create new sub-level if one doesn't exist
                    if (!list) {
                        list = document.createElement(this.options.listNodeName);
                        list.style.paddingLeft = this.options.threshold + PX;
                        list.appendChild(this._placeholder);
                        previous.appendChild(list);
                        // this.setParent(previous);
                    }
                    else {
                        // else append to next level up
                        list = previous.querySelector(":scope > " + this.options.listNodeName);
                        list.appendChild(this._placeholder);
                    }
                }
            }
            // decrease horizontal level
            if (this._mouse.distX < 0) {
                // we can't decrease a level if an item preceeds the current one
                var /** @type {?} */ next = document.querySelector("." + this.options.placeClass + " + " + this.options.itemNodeName);
                var /** @type {?} */ parentElement = this._placeholder.parentElement;
                if (!next && parentElement) {
                    var /** @type {?} */ closestItem = _closest(this._placeholder, this.options.itemNodeName);
                    if (closestItem) {
                        parentElement.removeChild(this._placeholder);
                        _insertAfter(this._placeholder, closestItem);
                    }
                }
            }
        }
        if (!pointEl.classList.contains('nestable-item-mask')) {
            return;
        }
        // find root list of item under cursor
        var /** @type {?} */ pointElRoot = _closest(this.pointEl, "." + this.options.rootClass), /** @type {?} */ isNewRoot = pointElRoot
            ? this.dragRootEl.dataset['nestable-id'] !==
                pointElRoot.dataset['nestable-id']
            : false;
        /**
         * move vertical
         */
        if (!this._mouse.dirAx || isNewRoot) {
            // check if groups match if dragging over new root
            if (isNewRoot &&
                this.options.group !== pointElRoot.dataset['nestable-group']) {
                return;
            }
            // check depth limit
            depth =
                this.dragDepth -
                    1 +
                    _getParents(this.pointEl, this.el.nativeElement.querySelector(this.options.listNodeName)).length;
            if (depth > this.options.maxDepth) {
                return;
            }
            var /** @type {?} */ before = event.pageY <
                _offset(this.pointEl).top + this.pointEl.clientHeight / 2;
            var /** @type {?} */ placeholderParent = this._placeholder.parentNode;
            // get point element depth
            var /** @type {?} */ pointRelativeDepth = void 0;
            pointRelativeDepth = _getParents(this.pointEl, this.el.nativeElement.querySelector(this.options.listNodeName)).length;
            if (this.options.fixedDepth) {
                if (pointRelativeDepth === this.relativeDepth - 1) {
                    var /** @type {?} */ childList = this.pointEl.querySelector(this.options.listNodeName);
                    if (!childList.children.length) {
                        childList.appendChild(this._placeholder);
                    }
                }
                else if (pointRelativeDepth === this.relativeDepth) {
                    if (before) {
                        this.pointEl.parentElement.insertBefore(this._placeholder, this.pointEl);
                    }
                    else {
                        _insertAfter(this._placeholder, this.pointEl);
                    }
                    if (Array.prototype.indexOf.call(this.pointEl.parentElement.children, this.pointEl) ===
                        this.pointEl.parentElement.children.length - 1) {
                        _insertAfter(this._placeholder, this.pointEl);
                    }
                }
            }
            else if (before) {
                this.pointEl.parentElement.insertBefore(this._placeholder, this.pointEl);
            }
            else {
                _insertAfter(this._placeholder, this.pointEl);
            }
        }
    };
    /**
     * @return {?}
     */
    NestableComponent.prototype.reset = function () {
        var /** @type {?} */ keys = Object.keys(this._mouse);
        for (var _i = 0, keys_1 = keys; _i < keys_1.length; _i++) {
            var key = keys_1[_i];
            this._mouse[key] = 0;
        }
        this._itemId = 0;
        this.moving = false;
        this.dragEl = null;
        this.dragRootEl = null;
        this.dragDepth = 0;
        this.relativeDepth = 0;
        this.hasNewRoot = false;
        this.pointEl = null;
    };
    /**
     * @param {?} event
     * @param {?} item
     * @param {?} parentList
     * @return {?}
     */
    NestableComponent.prototype.dragStartFromItem = function (event, item, parentList) {
        if (!this._registerHandleDirective) {
            this.dragStart(event, item, parentList);
        }
    };
    /**
     * @param {?} event
     * @param {?} item
     * @param {?} parentList
     * @return {?}
     */
    NestableComponent.prototype.dragStart = function (event, item, parentList) {
        this._oldListLength = this.list.length;
        if (!this.options.disableDrag) {
            event.stopPropagation();
            event.preventDefault();
            if (event.originalEvent) {
                event = event.originalEvent;
            }
            // allow only first mouse button
            if (event.type.indexOf('mouse') === 0) {
                if (event.button !== 0) {
                    return;
                }
            }
            else {
                if (event.touches.length !== 1) {
                    return;
                }
            }
            this.ref.detach();
            this._dragIndex = parentList.indexOf(item);
            this.dragModel = parentList.splice(parentList.indexOf(item), 1)[0];
            var /** @type {?} */ dragItem = _closest(event.target, this.options.itemNodeName);
            if (dragItem === null) {
                return;
            }
            this._parentDragId = Number.parseInt(dragItem.parentElement.parentElement.id);
            var /** @type {?} */ dragRect = dragItem.getBoundingClientRect();
            this._showMasks();
            this._createDragClone(event, dragItem);
            this.renderer.setStyle(this.dragEl, 'width', dragRect.width + PX);
            this._createPlaceholder(event, dragItem);
            this.renderer.setStyle(this._placeholder, 'height', dragRect.height + PX);
            this._calculateDepth();
            this.drag.emit({
                originalEvent: event,
                item: item
            });
            this._cancelMouseup = this.renderer.listen(document, 'mouseup', this.dragStop.bind(this));
            this._cancelMousemove = this.renderer.listen(document, 'mousemove', this.dragMove.bind(this));
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    NestableComponent.prototype.dragStop = function (event) {
        this._cancelMouseup();
        this._cancelMousemove();
        this._hideMasks();
        if (this.dragEl) {
            var /** @type {?} */ draggedId = Number.parseInt(this.dragEl.firstElementChild.id);
            var /** @type {?} */ placeholderContainer = _closest(this._placeholder, this.options.itemNodeName);
            var /** @type {?} */ changedElementPosition = this._dragIndex !==
                Array.prototype.indexOf.call(this._placeholder.parentElement.children, this._placeholder);
            var /** @type {?} */ index = Array.prototype.indexOf.call(this._placeholder.parentElement.children, this._placeholder);
            if ((this._dragIndex === index) && (this._oldListLength === this.list.length)) {
                changedElementPosition = true;
            }
            // placeholder in root
            if (placeholderContainer === null) {
                this.list.splice(Array.prototype.indexOf.call(this._placeholder.parentElement.children, this._placeholder), 0, Object.assign({}, this.dragModel));
            }
            else {
                // palceholder nested
                placeholderContainer = _findObjectInTree(this.list, Number.parseInt(placeholderContainer.id));
                if (!placeholderContainer.children) {
                    placeholderContainer.children = [];
                    placeholderContainer.children.push(Object.assign({}, this.dragModel));
                }
                else {
                    placeholderContainer.children.splice(Array.prototype.indexOf.call(this._placeholder.parentElement.children, this._placeholder), 0, Object.assign({}, this.dragModel));
                }
                if (index === this._dragIndex) {
                    changedElementPosition = false;
                }
                if (!changedElementPosition) {
                    changedElementPosition =
                        placeholderContainer['$$id'] !== this._parentDragId;
                }
            }
            this._placeholder.parentElement.removeChild(this._placeholder);
            this.dragEl.parentNode.removeChild(this.dragEl);
            this.dragEl.remove();
            this.reset();
            this.listChange.emit(this.list);
            this.drop.emit({
                originalEvent: event,
                destination: placeholderContainer,
                item: this.dragModel,
                changedElementPosition: changedElementPosition
            });
            this.ref.reattach();
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    NestableComponent.prototype.dragMove = function (event) {
        if (this.dragEl) {
            event.preventDefault();
            if (event.originalEvent) {
                event = event.originalEvent;
            }
            this._move(event.type.indexOf('mouse') === 0 ? event : event.touches[0]);
        }
    };
    /**
     * @return {?}
     */
    NestableComponent.prototype.expandAll = function () {
        _traverseChildren(this._list, function (item) {
            item['$$expanded'] = true;
        });
        this.ref.markForCheck();
    };
    /**
     * @return {?}
     */
    NestableComponent.prototype.collapseAll = function () {
        _traverseChildren(this._list, function (item) {
            item['$$expanded'] = false;
        });
        this.ref.markForCheck();
    };
    return NestableComponent;
}());
NestableComponent.decorators = [
    { type: core.Component, args: [{
                selector: 'ngx-nestable',
                template: "\n    <ng-container *ngTemplateOutlet=\"nestableList; context:{nestable: list, depth: 0}\">\n    </ng-container>\n\n    <ng-template #nestableList let-nestable=\"nestable\" let-depth=\"depth\">\n      <ul [style.padding-left.px]=\"depth ? options.threshold : 0\"\n        [class]=\"options.listClass\">\n        <li [class]=\"options.itemClass\" [id]=\"item.$id\" *ngFor=\"let item of nestable\">\n\n          <ng-container *ngTemplateOutlet=\"nestableItem; context:{nestable: nestable, item: item, depth: depth}\">\n          </ng-container>\n\n          <ng-container *ngIf=\"item.children && item.$expanded\">\n            <ng-container *ngTemplateOutlet=\"nestableList; context:{nestable: item.children, depth: depth + 1}\">\n            </ng-container>\n          </ng-container>\n\n          <ul [class]=\"options.listClass\" [style.padding-left.px]=\"options.threshold\"></ul>\n        </li>\n      </ul>\n    </ng-template>\n\n    <ng-template #nestableItem let-parentList=\"nestable\" let-item=\"item\" let-depth=\"depth\">\n      <div class=\"nestable-item-container mat-list-item\" (mousedown)=\"dragStartFromItem($event, item, parentList)\">\n        <ng-container *ngTemplateOutlet=\"template; context:{$implicit: {item: item, parentList: parentList}, depth: depth}\">\n        </ng-container>\n\n        <div class=\"nestable-item-mask\"></div>\n      </div>\n    </ng-template>\n  ",
                styles: ["\n    @import url(\"https://fonts.googleapis.com/css?family=Roboto\");\n    ul .dd-placeholder {\n      margin: 5px 0;\n      padding: 0;\n      min-height: 30px;\n      background: #f2fbff;\n      border: 1px dashed #b6bcbf;\n      -webkit-box-sizing: border-box;\n              box-sizing: border-box; }\n\n    ul li .nestable-item-mask {\n      display: none;\n      position: absolute;\n      top: 0;\n      bottom: 0;\n      right: 0;\n      left: 0;\n      z-index: 9998; }\n\n    ul li .nestable-expand-button {\n      display: block;\n      position: relative;\n      cursor: pointer;\n      float: left;\n      width: 25px;\n      height: 14px;\n      padding: 0;\n      white-space: nowrap;\n      overflow: hidden;\n      border: 0;\n      background: transparent;\n      font-size: 18px;\n      line-height: 1;\n      text-align: center;\n      font-weight: bold;\n      outline: 0; }\n\n    ul li .nestable-item-container {\n      position: relative;\n      display: -webkit-box;\n      display: -ms-flexbox;\n      display: flex;\n      -webkit-box-orient: horizontal;\n      -webkit-box-direction: normal;\n          -ms-flex-direction: row;\n              flex-direction: row;\n      -webkit-box-align: center;\n          -ms-flex-align: center;\n              align-items: center;\n      color: rgba(0, 0, 0, 0.87);\n      min-height: 32px;\n      font-size: 16px;\n      font-family: 'Roboto', sans-serif;\n      cursor: pointer;\n      outline: 0;\n      margin-bottom: 2px;\n      padding-left: 8px; }\n\n    ul li .nestable-item-container:hover {\n      background: rgba(0, 0, 0, 0.04); }\n\n    ol, ul {\n      list-style: none; }\n  "],
                encapsulation: core.ViewEncapsulation.None,
                changeDetection: core.ChangeDetectionStrategy.OnPush
            },] },
];
/**
 * @nocollapse
 */
NestableComponent.ctorParameters = function () { return [
    { type: core.ChangeDetectorRef, },
    { type: core.Renderer2, },
    { type: core.ElementRef, },
    { type: core.NgZone, },
]; };
NestableComponent.propDecorators = {
    'listChange': [{ type: core.Output },],
    'drop': [{ type: core.Output },],
    'drag': [{ type: core.Output },],
    'disclosure': [{ type: core.Output },],
    'template': [{ type: core.Input },],
    'options': [{ type: core.Input },],
    'disableDrag': [{ type: core.Input },],
    'list': [{ type: core.Input },],
};
var NestableDragHandleDirective = (function () {
    /**
     * @param {?} _el
     */
    function NestableDragHandleDirective(_el) {
        this._el = _el;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    NestableDragHandleDirective.prototype.onMouseDown = function (event) {
        var /** @type {?} */ detail = {
            param: this.ngxNestableDragHandle,
            event: event
        };
        this._el.nativeElement.dispatchEvent(new CustomEvent(DRAG_START, { bubbles: true, detail: detail }));
    };
    /**
     * @return {?}
     */
    NestableDragHandleDirective.prototype.ngOnInit = function () {
        this._el.nativeElement.dispatchEvent(new CustomEvent(REGISTER_HANDLE, { bubbles: true, detail: this.ngxNestableDragHandle }));
    };
    return NestableDragHandleDirective;
}());
NestableDragHandleDirective.decorators = [
    { type: core.Directive, args: [{
                selector: '[ngxNestableDragHandle]'
            },] },
];
/**
 * @nocollapse
 */
NestableDragHandleDirective.ctorParameters = function () { return [
    { type: core.ElementRef, },
]; };
NestableDragHandleDirective.propDecorators = {
    'ngxNestableDragHandle': [{ type: core.Input },],
    'onMouseDown': [{ type: core.HostListener, args: ['mousedown', ['$event'],] },],
};
var NestableExpandCollapseDirective = (function () {
    /**
     * @param {?} _el
     */
    function NestableExpandCollapseDirective(_el) {
        this._el = _el;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    NestableExpandCollapseDirective.prototype.onMouseDown = function (event) {
        event.stopPropagation();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    NestableExpandCollapseDirective.prototype.onClick = function (event) {
        this.ngxNestableExpandCollapse.item['$$expanded'] = !this.ngxNestableExpandCollapse.item['$$expanded'];
        this._el.nativeElement.dispatchEvent(new CustomEvent(EXPAND_COLLAPSE, {
            bubbles: true,
            detail: this.ngxNestableExpandCollapse
        }));
    };
    return NestableExpandCollapseDirective;
}());
NestableExpandCollapseDirective.decorators = [
    { type: core.Directive, args: [{
                selector: '[ngxNestableExpandCollapse]'
            },] },
];
/**
 * @nocollapse
 */
NestableExpandCollapseDirective.ctorParameters = function () { return [
    { type: core.ElementRef, },
]; };
NestableExpandCollapseDirective.propDecorators = {
    'ngxNestableExpandCollapse': [{ type: core.Input },],
    'onMouseDown': [{ type: core.HostListener, args: ['mousedown', ['$event'],] },],
    'onClick': [{ type: core.HostListener, args: ['click', ['$event'],] },],
};
var NestableModule = (function () {
    function NestableModule() {
    }
    return NestableModule;
}());
NestableModule.decorators = [
    { type: core.NgModule, args: [{
                imports: [common.CommonModule],
                declarations: [
                    NestableComponent,
                    NestableDragHandleDirective,
                    NestableExpandCollapseDirective
                ],
                exports: [
                    NestableComponent,
                    NestableDragHandleDirective,
                    NestableExpandCollapseDirective
                ]
            },] },
];
/**
 * @nocollapse
 */
NestableModule.ctorParameters = function () { return []; };

exports.NestableComponent = NestableComponent;
exports.NestableModule = NestableModule;
exports.ɵa = NestableDragHandleDirective;
exports.ɵb = NestableExpandCollapseDirective;

Object.defineProperty(exports, '__esModule', { value: true });

})));
